oxidation/libparsec/crates/client-connection/src/authenticated_cmds.rs

```diff
@@ -0,0 +1,318 @@
+/// How we serialize the data before sending a request.
+pub const PARSEC_CONTENT_TYPE: &str = "application/msgpack";
+
+/// Factory that send commands in a authenticated context.
+pub struct AuthenticatedCmds {
+    /// HTTP Client that contain the basic configuration to communicate with the server.
+    client: Client,
+    url: Url,
+    user_id: String,
+    signing_key: SigningKey,
+}
+
+impl AuthenticatedCmds {
+    /// Create a new `AuthenticatedCmds`
+    pub fn new(
+        client: Client,
+        root_url: &str,
+        user_id: &[u8],
+        signing_key: SigningKey,
+    ) -> Result<Self, url::ParseError> {
+        let root_url = Url::parse(root_url)?;
+        let url = root_url.join(AUTHENTICATED_API_URI)?;
+        let user_id = base64::encode(user_id);
+
+        Ok(Self {
+            client,
+            url,
+            user_id,
+            signing_key,
+        })
+    }
+}
+
+macro_rules! impl_auth_cmds {
+    (
+        $(
+            $(#[$outer:meta])*
+            $name:ident($($key:ident: $type:ty),*)
+        )+
+    ) => {
+        $(
+            $(#[$outer])*
+            pub async fn $name(&self, $($key: $type),*) -> command_error::Result<authenticated_cmds::$name::Rep> {
+                let request_builder = self.client.post(self.url.clone());
+
+                let data = authenticated_cmds::$name::Req::new($($key),*).dump().map_err(|e| CommandError::Serialization(e.to_string()))?;
+
+                let req = prepare_request(request_builder, &self.signing_key, &self.user_id, data).send();
+                let resp = req.await?;
+                if resp.status() != reqwest::StatusCode::OK {
+                    return Err(CommandError::InvalidResponseStatus(resp.status(), resp));
+                }
+
+                let response_body = resp.bytes().await?;
+
+                authenticated_cmds::$name::Rep::load(&response_body).map_err(CommandError::Deserialization)
+            }
+        )+
+    };
+}
+
+/// Prepare a new request, the body will be added to the Request using [RequestBuilder::body]
+fn prepare_request(
+    request_builder: RequestBuilder,
+    signing_key: &SigningKey,
+    user_id: &str,
+    body: Vec<u8>,
+) -> RequestBuilder {
+    let request_builder = sign_request(request_builder, signing_key, user_id, &body);
+
+    let mut content_headers = HeaderMap::with_capacity(2);
+    content_headers.insert(CONTENT_TYPE, HeaderValue::from_static(PARSEC_CONTENT_TYPE));
+    content_headers.insert(
+        CONTENT_LENGTH,
+        HeaderValue::from_str(&body.len().to_string()).expect("numeric value are valid char"),
+    );
+
+    request_builder.headers(content_headers).body(body)
+}
+
+/// Sing a request by adding specific headers.
+fn sign_request(
+    request_builder: RequestBuilder,
+    signing_key: &SigningKey,
+    user_id: &str,
+    body: &[u8],
+) -> RequestBuilder {
+    use std::time::SystemTime;
+
+    let timestamp = SystemTime::now()
```

Comment:
Can you use chrono instead ? It's what we use in the rest of the codebase, and it allows to be much more explicit when doing the serialization (e.g. `my_data.to_rfc3339_opts(chrono::SecondsFormat::Secs, false)`)

Reply:
I can, but I've chosen not to, to limit the number of deps

Reply:
I find the discussion here being a sub-part of https://github.com/Scille/parsec-cloud/pull/2460#discussion_r902504277

So I prefer to resolve the main discussion before continuing this one

Reply:
outdated because of https://github.com/Scille/parsec-cloud/pull/2460#discussion_r905924039 and 0d10c11